### PR TITLE
[sig-storage] Add custom timeouts in E2E tests

### PR DESCRIPTION
**Pull request by bertinatto**
**What type of PR is this?**

/kind bug
/kind flake

**What this PR does / why we need it**:

Some storage backends can be slower or faster when it comes to certain operations like attaching or detaching a disk. In other words, what is considered to be a normal operation duration to some backends might not be OK for others.

This frequently results in flakes in certain clouds. This PR tries to improve this situation by allowing storage drivers (from E2E tests) to specify their own timeouts.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This PR is not ready for review yet.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/e2e/storage/drivers/in_tree.go [hunk 1]
```diff
@@ -1605,6 +1606,12 @@ func (a *azureDiskDriver) PrepareTest(f *framework.Framework) (*testsuites.PerTe
 	}, func() {}
 }
 
+func (a *azureDiskDriver) GetTimeouts() framework.TimeoutContext {
+	t := framework.NewTimeoutContextWithDefaults()
+	t.PodStart = time.Minute * 10
+	return t
+}
```
Comment (bertinatto) [inline test/e2e/storage/drivers/in_tree.go:1613]: This (custom timeouts in Azure Disk) is just an example of how to add custom timeouts to an in-tree driver. I plan to remove it before merging this PR.

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -71,9 +71,23 @@ import (
...
+
+	// TODO(justinsb): Avoid hardcoding this.
+	awsMasterIP = "172.20.0.9"
+
+	// AllContainers specifies that all containers be visited
+	// Copied from pkg/api/v1/pod to avoid pulling extra dependencies
+	AllContainers = InitContainers | Containers | EphemeralContainers
+)
+
+// DEPRECATED constants. Use the timeouts in framework.Framework instead.
```
Comment (bertinatto) [inline test/e2e/framework/util.go:86]: Ideally we'd remove these constants from here in favor of the values in framework.Framework.Timeouts. However, these consts are used in **too many places** throughout the test suite and changing all at once would make this PR too big (I tried that).

So the approach I found more convenient was to deprecate the consts in `util.go` and and remove them over time, once there isn't any test using them any more.

File: test/e2e/framework/timeouts.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	PodGet time.Duration
+
+	// PodEvent is how much we wait for a pod event to occur.
+	PodEvent time.Duration
+
+	// ServiceStart is how long to wait for a service endpoint to be resolvable.
+	ServiceStart time.Duration
+
+	// Poll is how often to Poll pods, nodes and claims.
+	Poll time.Duration
```
Comment (jsafrane) [inline test/e2e/framework/timeouts.go:47]: Does this need to be configurable? IMO it's about the frequency and IMO we can have one that makes everyone happy - once per 2s should be OK.

File: test/e2e/framework/timeouts.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	PVBinding time.Duration
+
+	// ClaimBindingTimeout is how long claims have to become bound.
+	ClaimBinding time.Duration
+
+	// PVReclaimingTimeout is how long PVs have to beome reclaimed.
+	PVReclaiming time.Duration
+
+	// PVDeletingTimeout is how long PVs have to become deleted.
+	PVDeleting time.Duration
```
Comment (jsafrane) [inline test/e2e/framework/timeouts.go:101]: These either need a better name or be merged with some timeouts above (ClaimBinding -> ClaimProvisionShort?, PVDeleting -> rename to PVDelete,...)

File: test/e2e/framework/timeouts.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import "time"
+
+// TimeoutContext contains timeout settings for several actions.
+type TimeoutContext struct {
```
Comment (jsafrane) [inline test/e2e/framework/timeouts.go:22]: There is lot of timeouts in this struct that are never used in this PR. Does it make sense to pre-allocate them for subsequent, non-storage PRs?

File: test/e2e/framework/timeouts.go (hunk 3 above)
Comment (bertinatto) [reply test/e2e/framework/timeouts.go]: Removed

File: test/e2e/framework/timeouts.go (hunk 4 above)
Comment (bertinatto) [reply test/e2e/framework/timeouts.go:95]: These are copies of the original values [here](https://github.com/kubernetes/kubernetes/pull/96042/commits/b37e7b701afa1aeb1477b2b95bafe8fdf24e06d9#diff-f9cf7d59f38dc004bb86a00b169c989bf93d49636bdf254bf38909a3d37816e3L41-L52), but I'm OK with changing the names.

File: test/e2e/framework/timeouts.go (hunk 5 above)
Comment (bertinatto) [reply test/e2e/framework/timeouts.go:22]: My original idea was to:

1. Move all timeouts to this struct
2. Deprecate the original values used here: https://github.com/kubernetes/kubernetes/blob/f975a2d485ca7ce559d437ed6a1b07111bf1404d/test/e2e/framework/util.go#L86
3. Use the values from this struct _in all e2e tests_, not only storage tests.

However, when I got to step 3 above I noticed that this PR would become excessively big, so I started with storage tests only and hoped that eventually the other tests would use these values instead (since the values in utils.go are deprecated).

Does this make sense or should I just add to this struct the timeouts I'm going to use in storage tests?

The reason I wanted to do this is because currently timeout values are scattered around framework package: utils.go, pv/pv.go etc. So I'd like this file to be the source of truth for all values.

File: test/e2e/storage/external/external.go [hunk 6]
```diff
@@ -302,6 +309,34 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
 	return testsuites.GetStorageClass(sc.Provisioner, sc.Parameters, sc.VolumeBindingMode, f.Namespace.Name, "e2e-sc")
 }
 
+func (d *driverDefinition) GetTimeouts() *framework.TimeoutContext {
+	timeouts := framework.NewTimeoutContextWithDefaults()
+	if d.Timeouts != nil {
```
Comment (saikat-royc) [inline test/e2e/storage/external/external.go:314]: Can we instead check if d.Timeouts == nil { return timeouts} and avoid the indent.

File: test/e2e/framework/timeouts.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	PVReclaiming time.Duration
+
+	// PVDeletingTimeout is how long PVs have to become deleted.
+	PVDeleting time.Duration
+}
+
+// NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+func NewTimeoutContextWithDefaults() *TimeoutContext {
+	return &TimeoutContext{
+		PodList:                 1 * time.Minute,
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:101]: Don't think we need custom timeouts for listing/getting resources. Thoughts?

File: test/e2e/framework/timeouts.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	// PVReclaimingTimeout is how long PVs have to beome reclaimed.
+	PVReclaiming time.Duration
+
+	// PVDeletingTimeout is how long PVs have to become deleted.
+	PVDeleting time.Duration
+}
+
+// NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+func NewTimeoutContextWithDefaults() *TimeoutContext {
+	return &TimeoutContext{
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:100]: Can we create const for the default timeout values in this file? and then use them here.
e.g PVBindingTimeout etc

File: test/e2e/storage/external/external.go [hunk 9]
```diff
@@ -128,6 +131,8 @@ type driverDefinition struct {
 	// Can be left empty. Most drivers should not need this and instead
 	// use topology to ensure that pods land on the right node(s).
 	ClientNodeName string
+
+	Timeouts map[string]string
```
Comment (saikat-royc) [inline test/e2e/storage/external/external.go:135]: Can you put some comments for this map

File: test/e2e/storage/external/external.go [hunk 10]
```diff
@@ -302,6 +309,34 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
...
+func (d *driverDefinition) GetTimeouts() *framework.TimeoutContext {
+	timeouts := framework.NewTimeoutContextWithDefaults()
+	if d.Timeouts != nil {
+		c := make(map[string]time.Duration)
+		for k, v := range d.Timeouts {
+			duration, err := time.ParseDuration(v)
+			if err != nil {
+				// We can't use ExpectNoError() because his method can be called out of an It(),
+				// so we simply log the error and return the default timeouts.
+				klog.Errorf("Could not parse duration: %v", err)
```
Comment (saikat-royc) [inline test/e2e/storage/external/external.go:321]: Can you also specify the key here for debugging: "Could not parse duration %v for key %s"

File: test/e2e/framework/timeouts.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+		SingleCall:              5 * time.Minute,
+		NodeReadyInitial:        20 * time.Second,
+		PodReadyBefore:          5 * time.Minute,
+		ClaimProvisionShort:     1 * time.Minute,
+		ClaimProvision:          5 * time.Minute,
+		RestartNodeReadyAgain:   5 * time.Minute,
+		RestartPodReadyAgain:    5 * time.Minute,
+		SnapshotCreate:          5 * time.Minute,
+		SnapshotDelete:          5 * time.Minute,
+		// pv.go
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:118]: align the indentation values

**Comment by saikat-royc** [comment]
@bertinatto Thanks for making the changes. Can you rebase the patch and add the changes to new test suites like [fsgroupchangepolicy.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/fsgroupchangepolicy.go)

File: test/e2e/framework/timeouts.go (hunk 5 above)
Comment (saikat-royc) [reply test/e2e/framework/timeouts.go:22]: Keeping the timeouts in a single place (timeouts.go) makes the code readable. I think we can start with the timeouts we are using in storage tests.

File: test/e2e/framework/timeouts.go (hunk 4 above)
Comment (saikat-royc) [reply test/e2e/framework/timeouts.go:95]: Let's create the new variables with proper names as Jan suggested. The old ones will eventually be deprecated.

File: test/e2e/framework/pv/pv.go [hunk 12]
```diff
@@ -38,18 +39,6 @@ const (
...
-	PVBindingTimeout = 3 * time.Minute
-
-	// ClaimBindingTimeout is how long claims have to become bound.
-	ClaimBindingTimeout = 3 * time.Minute
-
-	// PVReclaimingTimeout is how long PVs have to beome reclaimed.
-	PVReclaimingTimeout = 3 * time.Minute
-
-	// PVDeletingTimeout is how long PVs have to become deleted.
-	PVDeletingTimeout = 3 * time.Minute
```
Comment (bertinatto) [inline test/e2e/framework/pv/pv.go:51]: These values are not used anywhere else, so I removed instead of deprecating them.

File: test/e2e/framework/timeouts.go (hunk 5 above)
Comment (bertinatto) [reply test/e2e/framework/timeouts.go:37]: Sounds good. Removed everything that was not used.

File: test/e2e/framework/timeouts.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+package framework
+
+import "time"
+
+const (
+	// Default timeouts to be used in TimeoutContext
+	podStartTimeout            = 5 * time.Minute
+	podStartShortTimeout       = 2 * time.Minute
+	podDeleteTimeout           = 5 * time.Minute
+	claimProvisionTimeout      = 5 * time.Minute
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:26]: minor nit: its more readable to keep the timeouts for a given resource together. e.g claimProvisionTimeout, claimProvisionShortTimeout, claimBoundTimeout

File: test/e2e/framework/timeouts.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+// TimeoutContext contains timeout settings for several actions.
+type TimeoutContext struct {
+	// PodStart is how long to wait for the pod to be started.
+	PodStart time.Duration
+
+	// PodStartShort is same as `PodStart` to wait for the pod to be started, but shorter.
+	// Use it case by case when we are sure pod start will not be delayed.
+	// minutes by slow docker pulls or something else.
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:43]: please revisit this sentence. "minutes by slow docker pulls or something else"

File: test/e2e/framework/timeouts.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+	PVDelete time.Duration
+}
+
+// NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+func NewTimeoutContextWithDefaults() *TimeoutContext {
+	return &TimeoutContext{
+		PodStart:            podStartTimeout,
+		PodStartShort:       podStartShortTimeout,
+		PodDelete:           podDeleteTimeout,
+		ClaimProvision:      claimProvisionTimeout,
```
Comment (saikat-royc) [inline test/e2e/framework/timeouts.go:81]: same comment as above about grouping resources together.

File: test/e2e/storage/csi_mock_volume.go [hunk 16]
```diff
@@ -383,7 +384,7 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			}.AsSelector().String()
 			msg := "AttachVolume.Attach failed for volume"
 
-			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, framework.PodStartTimeout)
+			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, f.Timeouts.PodStart)
 			if err != nil {
 				podErr := e2epod.WaitTimeoutForPodRunningInNamespace(m.cs, pod.Name, pod.Namespace, 10*time.Second)
```
Comment (saikat-royc) [inline test/e2e/storage/csi_mock_volume.go:389]: 10*time.Second replace this with podStartShortTimeout?

File: test/e2e/storage/testsuites/provisioning.go [hunk 17]
```diff
@@ -45,6 +45,7 @@ import (
 // Not all parameters are used by all tests.
```
Comment (saikat-royc) [inline test/e2e/storage/testsuites/provisioning.go:45]: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/provisioning.go#L388,
```framework.ExpectNoError(e2epv.WaitForPersistentVolumeDeleted(client, pv.Name, 5*time.Second, 20*time.Minute))```
there is a wait for pv delete step that has 20 mins timeout not sure why. This needs to be revisited, perhaps in a different PR.

File: test/e2e/framework/volume/fixtures.go [hunk 18]
```diff
@@ -441,11 +441,11 @@ func runVolumeTesterPod(client clientset.Interface, config TestConfig, podSuffix
 	if slow {
 		err = e2epod.WaitForPodRunningInNamespaceSlow(client, clientPod.Name, clientPod.Namespace)
```
Comment (saikat-royc) [inline test/e2e/framework/volume/fixtures.go:442]: this function indirectly uses https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go where the pod specific timeouts are defined. Those timeouts should also eventually move to the timeouts.go. we can put a todo here?

File: test/e2e/storage/external/external.go [hunk 19]
```diff
@@ -302,6 +312,42 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
...
+	// Convert the temporary map holding the custom timeouts to JSON
+	t, err := json.Marshal(c)
+	if err != nil {
+		klog.Errorf("Could not marshal timeouts: %v", err)
+		return timeouts
+	}
+
+	// Override the default timeouts with the custom ones
+	err = json.Unmarshal(t, &timeouts)
+	if err != nil {
```
Comment (jingxu97) [inline test/e2e/storage/external/external.go:343]: just want to make sure at this point, the returned timeouts will still be the default value?
also in the log, maybe mention returning default timeout?

File: test/e2e/storage/testsuites/testdriver.go [hunk 20]
```diff
@@ -129,6 +129,19 @@ type SnapshottableTestDriver interface {
 	GetSnapshotClass(config *PerTestConfig) *unstructured.Unstructured
 }
 
+// CustomTimeoutsTestDriver represents an interface fo a TestDriver that supports custom timeouts.
+type CustomTimeoutsTestDriver interface {
```
Comment (jingxu97) [inline test/e2e/storage/testsuites/testdriver.go:133]: just a suggestion: TestDriverWithTimeout?

File: test/e2e/storage/testsuites/testdriver.go [hunk 21]
```diff
@@ -129,6 +129,19 @@ type SnapshottableTestDriver interface {
...
+	TestDriver
+	GetTimeouts() *framework.TimeoutContext
+}
+
+func getCustomTimeouts(driver TestDriver) *framework.TimeoutContext {
```
Comment (jingxu97) [inline test/e2e/storage/testsuites/testdriver.go:138]: getDriverTimeouts?

**Comment by jingxu97** [comment]
thanks for the PR! @bertinatto just a few small comments. overall lgtm.

cc @Jiawei0227

File: test/e2e/storage/external/external.go [hunk 22]
```diff
@@ -302,6 +312,42 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
...
 }
 
+func (d *driverDefinition) GetTimeouts() *framework.TimeoutContext {
+	timeouts := framework.NewTimeoutContextWithDefaults()
+	if d.Timeouts == nil {
+		return timeouts
+	}
+
+	// Use a temporary map to hold the timeouts specified in the manifest file
+	c := make(map[string]time.Duration)
```
Comment (saikat-royc) [inline test/e2e/storage/external/external.go:322]: The map should be created with default values. And overwrite the values for the keys provided in the driver config.

Comment (bertinatto) [reply test/e2e/storage/external/external.go:322]: The `timeouts` struct above contains the default timeouts. This map only contains the custom values provided by the driver config.

Note that down below map is then merged with the struct, overwriting the default values with the custom ones.

Comment (saikat-royc) [reply test/e2e/storage/external/external.go:322]: Ah I see. thanks for pointing it out. missed it.

File: test/e2e/storage/external/external.go (hunk 19 above)
Comment (bertinatto) [reply test/e2e/storage/external/external.go:343]: Right, `timeouts` contains the default values at this point.

Extended log message...

File: test/e2e/storage/testsuites/testdriver.go (hunk 20 above)
Comment (bertinatto) [reply test/e2e/storage/testsuites/testdriver.go:133]: That sounds better indeed, but it'd differ from the other interfaces (they are suffixed with `TestDriver`)

File: test/e2e/storage/testsuites/testdriver.go (hunk 21 above)
Comment (bertinatto) [reply test/e2e/storage/testsuites/testdriver.go]: Renamed

File: test/e2e/storage/testsuites/provisioning.go (hunk 17 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go:45]: I didn't add the comment, I just moved it to a new place :-)

I had to trace back quite a bit to where this comment was originally added: https://github.com/kubernetes/kubernetes/commit/46fb4172999bebcf822a2953e1783ed0428adc70

I think the 20 mins is to also account for the time period that it may take to unmount and detach the volume. 20 minutes seems quite long and could probably be part of the customizable timeouts?

File: test/e2e/storage/csi_mock_volume.go [hunk 23]
```diff
@@ -383,7 +384,7 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			}.AsSelector().String()
 			msg := "AttachVolume.Attach failed for volume"
 
-			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, framework.PodStartTimeout)
+			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, f.Timeouts.PodStart)
```
Comment (msau42) [inline test/e2e/storage/csi_mock_volume.go:387]: Hm so if a plugin decides to set a non-default timeout, then that will impact the timeout for non driver specific tests?

Is that what we want?

Comment (saikat-royc) [reply test/e2e/storage/csi_mock_volume.go:387]: can you point to an e.g for non-driver specific test?

Comment (msau42) [reply test/e2e/storage/csi_mock_volume.go:387]: I was thinking specifically about this csi mock driver test. If you run the csi mock driver test along with a cloud plugin that takes minutes to provision/attach, etc, then that increases the timeout on the mock driver tests, which aren't expected to take very long, and could mask other problems in the system.

**Comment by saikat-royc** [comment]
Looking at one more example of subpath.go tests, the function [testReadFile](
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L682:6), uses [TestContainerOutput](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L687). The calls lands finally to [MatchContainerOutput](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L803:21). This means for a driver which supports waitforfirstcustomer  (delayed binding), we are not able to control the pod start time, since the start time would encompass the volume provisioning time too.

File: test/e2e/storage/csi_mock_volume.go (hunk 23 above)
Comment (saikat-royc) [reply test/e2e/storage/csi_mock_volume.go:387]: I see, this seems to be a csi mock plugin specific problem. For the testsuites, they are run in the context of a driver (intree, csi or external). We can remove the non-default timeouts for the mock driver. thoughts?

Comment (msau42) [reply test/e2e/storage/csi_mock_volume.go:387]: I guess right now these timeouts are only configurable via external storage interface or a custom ginkgo suite, so it's unlikely for someone to be running both csi mock and external storage tests in the same run. I would be more concerned if we want to extend this to all e2es in the future.

Comment (bertinatto) [reply test/e2e/storage/csi_mock_volume.go:387]: Notice that this csi mock driver test is not using custom timeouts, it's using the default ones:

https://github.com/kubernetes/kubernetes/pull/96042/files/f594e0f7737a0d281e9b08ddf1b1ec770755ee6a#diff-ffda3f0c41f167e7deedad89b23cf9b89d0e9805f910c81e075baa19d667f21fR137

Even if it was using the custom timeouts, it would get the values returned by the mock driver's `GetTimeouts()` method, as in:

```diff
diff --git a/test/e2e/storage/drivers/csi.go b/test/e2e/storage/drivers/csi.go
index 03ec35e557c..d4956d47389 100644
--- a/test/e2e/storage/drivers/csi.go
+++ b/test/e2e/storage/drivers/csi.go
@@ -278,6 +278,7 @@ type CSIMockDriverOpts struct {
 var _ testsuites.TestDriver = &mockCSIDriver{}
 var _ testsuites.DynamicPVTestDriver = &mockCSIDriver{}
 var _ testsuites.SnapshottableTestDriver = &mockCSIDriver{}
+var _ testsuites.CustomTimeoutsTestDriver = &mockCSIDriver{}

 // InitMockCSIDriver returns a mockCSIDriver that implements TestDriver interface
 func InitMockCSIDriver(driverOpts CSIMockDriverOpts) testsuites.TestDriver {
@@ -336,6 +337,12 @@ func InitMockCSIDriver(driverOpts CSIMockDriverOpts) testsuites.TestDriver {
        }
 }

+func (m *mockCSIDriver) GetTimeouts() *framework.TimeoutContext {
+       timeouts := framework.NewTimeoutContextWithDefaults()
+       timeouts.PodStart = time.Second * 5
+       return timeouts
+}
+
 func (m *mockCSIDriver) GetDriverInfo() *testsuites.DriverInfo {
        return &m.driverInfo
 }
```

Comment (msau42) [reply test/e2e/storage/csi_mock_volume.go:387]: Awesome, this addresses my concern for now thanks!

**Comment by msau42** [comment]
/approve

Will let @saikat-royc give lgtm

File: test/e2e/storage/pvc_protection.go [hunk 24]
```diff
@@ -18,12 +18,13 @@ package storage
...
 import (
 	"context"
+
 	"github.com/onsi/ginkgo"
 
 	"fmt"
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (saikat-royc) [inline test/e2e/storage/pvc_protection.go:27]: is this change needed?

File: test/e2e/storage/vsphere/vsphere_stress.go [hunk 25]
```diff
@@ -23,7 +23,7 @@ import (
 
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (saikat-royc) [inline test/e2e/storage/vsphere/vsphere_stress.go:26]: same comment here

**Comment by saikat-royc** [comment]
The changes look good to me with 2 minor comments. 

/lgtm

**Comment by saikat-royc** [comment]
@bertinatto  found some more instances of WaitForPodSuccessInNamespace() calls in the storage testsuites. e.g [subpath.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L1013).
some more in provisioning.go etc, where we need to use the configurable pod start

**Review by bertinatto — COMMENTED** [review]
```
@bertinatto found some more instances of WaitForPodSuccessInNamespace() calls in the storage testsuites. e.g subpath.go.
some more in provisioning.go etc, where we need to use the configurable pod start
```
@saikat-royc I think I've replaced all occurrences of  `WaitForPodSuccessInNamespace*` by `WaitForPodSuccessInNamespaceTimeout` in storage tests. PTAL

**Comment by saikat-royc** [comment]
The test `pull-kubernetes-e2e-gce-ubuntu-containerd` seems to be failing due to failure to pull image
```BackOff: Back-off pulling image "gcr.io/k8s-authenticated-test/agnhost:2.6"```

**Comment by bertinatto** [comment]
```
Kubernetes e2e suite: [sig-apps] ReplicationController should serve a basic image on each replica with a private image expand_more 	5m10s
Kubernetes e2e suite: [sig-apps] ReplicaSet should serve a basic image on each replica with a private image expand_more
```

The failure doesn't seem related to this PR.

/retest

**Comment by bertinatto** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/97002
